blockpresencemanager/blockpresencemanager.go

```diff
@@ -0,0 +1,109 @@
+package blockpresencemanager
+
+import (
+	"sync"
+
+	cid "github.com/ipfs/go-cid"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+)
+
+type BlockPresenceManager struct {
+	sync.RWMutex
+	presence map[cid.Cid]map[peer.ID]bool
```

Comment:
There will be a limited number of peer IDs so another option might be to store an index into an array of peer IDs instead of the peer IDs themselves

Reply:
That shouldn't matter. We should only have ~1 actual _copy_ of each peer ID in memory, `peer.ID` is just a pointer.

My concern is that we won't remove these until we disconnect, right? Ah, wait, do we remove these when we receive the blocks?

Reply:
This class maintains a table of which peers have sent HAVE or DONT_HAVE for each CID we want:
```
      peer A   peer B   peer C
CID1    HAVE     DONT     HAVE
CID2    DONT        ?     HAVE
CID3    HAVE        ?     ?
```

We remove them individually as we receive the blocks. We remove them all when the session ends (all blocks received or context closes).

The table will grow as more wants are requested, and as more peers are added to the session. So the speed at which it grows will depend mostly on the request pattern and on the block distribution in the network:
- if the DAG has a high branch out factor then the number of CIDs could grow large fairly quickly.
- if there are a lot of peers participating in the session (meaning there are a lot of peers in the network that have some of the blocks requested by the session) then there will be a lot of incoming HAVE / DONT_HAVE messages which could also cause the table to grow quickly

Once the session closes it will get cleaned up though, so if the session timeout is not too long it may not be an issue.